File: python/Day1.py
from collections import Counter
from typing import List, Tuple

def calculate_total_distance(left_list: List[int], right_list: List[int]) -> int:
    """
    Calculate the total distance between corresponding elements of two integer lists.
    Both lists are first sorted to pair smallest with smallest, etc.
    The distance is the sum of absolute differences of paired elements.

    Parameters:
        left_list (List[int]): The first list of integers.
        right_list (List[int]): The second list of integers.

    Returns:
        int: The sum of absolute differences between paired elements.
    """
    left_sorted = sorted(left_list)
    right_sorted = sorted(right_list)
    return sum(abs(l - r) for l, r in zip(left_sorted, right_sorted))

def calculate_similarity_score(left_list: List[int], right_list: List[int]) -> int:
    """
    Calculate a similarity score between two lists.
    The score is defined as the sum of the product of an element in `left_list` and
    its occurrences in `right_list`.

    Parameters:
        left_list (List[int]): The first list of integers.
        right_list (List[int]): The second list of integers.

    Returns:
        int: The calculated similarity score.
    """
    right_count = Counter(right_list)
    return sum(num * right_count[num] for num in left_list)

def read_lists_from_file(filename: str) -> Tuple[List[int], List[int]]:
    """
    Read pairs of integers from the given file and separate them into two lists.

    The file is expected to have lines with two integers per line: `left right`.

    Parameters:
        filename (str): The path to the input file.

    Returns:
        Tuple[List[int], List[int]]: Two lists, one containing all left integers and the other all right integers.
    """
    left_list, right_list = [], []
    with open(filename, 'r') as file:
        for line in file:
            left, right = map(int, line.split())
            left_list.append(left)
            right_list.append(right)
    return left_list, right_list

def main():
    """Execute Day 1 puzzle solution."""
    file_path = '../data/Day1.txt'
    left_list, right_list = read_lists_from_file(file_path)

    # Part 1
    total_distance = calculate_total_distance(left_list, right_list)
    print(f"Day 1 - Part 1: Total distance: {total_distance}")

    # Part 2
    similarity_score = calculate_similarity_score(left_list, right_list)
    print(f"Day 1 - Part 2: Similarity score: {similarity_score}")

if __name__ == "__main__":
    main()

File: C#/Day1.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Linq;

namespace AdventSolutions
{
    /// <summary>
    /// Day 1 solution translated to C#, handling calculation of total distance and similarity score.
    /// </summary>
    public static class Day1
    {
        /// <summary>
        /// Reads pairs of integers from a file into two lists.
        /// </summary>
        private static (List<int> LeftList, List<int> RightList) ReadListsFromFile(string filename)
        {
            var leftList = new List<int>();
            var rightList = new List<int>();

            foreach (var line in File.ReadLines(filename))
            {
                var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 2 && int.TryParse(parts[0], out int left) && int.TryParse(parts[1], out int right))
                {
                    leftList.Add(left);
                    rightList.Add(right);
                }
            }

            return (leftList, rightList);
        }

        /// <summary>
        /// Calculate the total distance between corresponding elements of two lists after sorting them.
        /// </summary>
        private static int CalculateTotalDistance(List<int> leftList, List<int> rightList)
        {
            var leftSorted = leftList.OrderBy(x => x).ToList();
            var rightSorted = rightList.OrderBy(x => x).ToList();

            return leftSorted.Zip(rightSorted, (l, r) => Math.Abs(l - r)).Sum();
        }

        /// <summary>
        /// Calculate similarity score defined as sum of (element * frequency_of_element_in_other_list).
        /// </summary>
        private static int CalculateSimilarityScore(List<int> leftList, List<int> rightList)
        {
            var rightCount = rightList.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());

            int similarityScore = 0;
            foreach (var num in leftList)
            {
                if (rightCount.TryGetValue(num, out int count))
                {
                    similarityScore += num * count;
                }
            }

            return similarityScore;
        }

        public static void Run()
        {
            string filePath = "../data/Day1.txt";
            if (!File.Exists(filePath))
            {
                Console.WriteLine("Day1 data file not found.");
                return;
            }

            var (leftList, rightList) = ReadListsFromFile(filePath);

            // Part 1
            int totalDistance = CalculateTotalDistance(leftList, rightList);
            Console.WriteLine($"Day 1 - Part 1: Total distance: {totalDistance}");

            // Part 2
            int similarityScore = CalculateSimilarityScore(leftList, rightList);
            Console.WriteLine($"Day 1 - Part 2: Similarity score: {similarityScore}");
        }
    }
}
